### ISSUE TITLE
Implement a randomization algorithm

### ISSUE THREAD

**Issue by connorkuehl** [connorkuehl/clang_randstruct#49]
It is likely we will be able to accomplish much of the "shuffling" with the `std` library. However, keep in mind things such as alignment, etc.

**Comment by connorkuehl** [comment]
Forgot to update this the other day, but we had a discussion regarding the possibility of fields that should be adjacent for alignment purposes otherwise we would require padding. Just a note to keep these things in mind. We brought up the topic of an initial "partitioning" pass for the fields so we can categorize and classify these prior to randomization.

**Comment by connorkuehl** [comment]
We just had a discussion around alignment constraints. If Clang performs padding to make sure every field is properly aligned, this becomes less difficult.

In the event that Clang will not pad for us when we randomize:
* We can organize the fields into "blocks" according to their sizes (8 byte, 4 byte, 2 byte, 1 byte)
* We can randomize within those blocks to keep them adjacent to one-another within the blocks
* Spit out the contents of the highest-sized block, followed by next-highest, etc.

In the event that Clang _will_ pad appropriately for us:
* Any form of randomization will suffice.

We need to be cognizant of things like bit vectors though.

**Comment by connorkuehl** [comment]
It looks like our offsets will need to be "padding aware" according to #68

**Comment by connorkuehl** [comment]
> It is important to note that the last member is padded with the number of bytes required so that the total size of the structure should be a multiple of the largest alignment of any structure member [citation needed]
- [Wikipedia on Data structure alignment](https://en.wikipedia.org/wiki/Data_structure_alignment#Typical_alignment_of_C_structs_on_x86)

Using offsets, if faced with a situation like:

```c
struct ohdear {
    char potato;
    int dragon;
    char cherry;
    short spud;
    char uhoh;
};
```

We could do:

```c
struct ohdear {
    char potato; // potato is aligned
    char pad0[3]; // offset dragon 4 bytes total
    int dragon; // dragon is aligned
    char cherry; // cherry is aligned
    char pad1; // offset 2 bytes so spud is aligned
    short spud;
    char uhoh; // uhoh is aligned
};
```

Largest data type = int = 4 bytes, so `ohdear` structures should be a multiple of 4.

`sizeof(struct ohdear)` = 1 + 3 + 4 + 1 + 1 + 2 + 1 = 13 bytes :skull: We need the size of this structure to be at least 16 in this case.

In this situation, padding would be added onto the end. Could we accomplish the same by adjusting the offset of the last field and just have a bigger gap? I am not entirely sure for all cases, but it seems like that could potentially malign the last field.

**Comment by connorkuehl** [comment]
> In this situation, padding would be added onto the end. Could we accomplish the same by adjusting the offset of the last field and just have a bigger gap? I am not entirely sure for all cases, but it seems like that could potentially malign the last field.

Okay, I think we've got some alignment/padding-aware offsetting happening as of commit: 78339c8ecdd4fcc3a38255c5f3398ecdc80ec11d

I did test it against the structure described in my last comment here, and it is clear that tail-padding is required.

**Comment by connorkuehl** [comment]
>  clear that tail-padding is required.

I also just realized how easy it is to pad after the last data member. 

If the structure is 120 bits instead of 128, the plugin can tell Clang that the structure is 128 bits to achieve "tail padding"

**Comment by connorkuehl** [comment]
Closed by #85

**Comment by connorkuehl** [comment]
Moving forward our algorithm will certainly need to be more sophisticated so that:

- We have adjacent regions of the structure that can fit into cache lines
- We do not cause bizarre memory regressions by randomizing in a naive way that results in huge amounts of padding being added

This work is being tracked in #83
